File: python_solutions/chapter_08_recursion_and_dynamic_programming/problem_08_03_magic_index.py
def magic_index(A, start=None, end=None):
    if start is None:
        start = 0
    if end is None:
        end = len(A) - 1
    if len(A) < 1 or end - start < 0:
        return None
    i = (end + start) // 2
    if i == A[i]:
        return i
    if i < A[i]:
        return magic_index(A, 0, i)
    return magic_index(A, i+1, end)

File: cpp_solutions/chapter_08_recursion_and_dynamic_programming/problem_08_03_magicIndex.cpp
#include "problem_08_03_magicIndex.h"

namespace chapter_08{
    int binarySearchMagicIndex(const std::vector<int>& array, int start, int end){
        int middle = (start + end) / 2;
        if(array.size() < 1 || end - start < 0){
            return -1;
        }
        if(middle == array[middle]){
            return middle;
        }
        if(array.size() == 1){
            return -1;
        }
        if (middle < array[middle]){
            return binarySearchMagicIndex(array, start, middle);
        } else {
            return binarySearchMagicIndex(array, middle + 1, end);
        }
    }

    int magicIndex(const std::vector<int>& array){
        return binarySearchMagicIndex(array, 0, array.size() -1);
    }
}
